snapcraft/internal/lifecycle/_status_cache.py

```diff
@@ -0,0 +1,174 @@
+            1. Hasn't yet run
+            2. Is dirty
+            3. Either (1) or (2) apply to any earlier steps in its lifecycle
+        """
+        if (not self.step_has_run(part, step) or
+                self.get_dirty_report(part, step) is not None):
+            return True
+
+        previous_step = step.previous_step()
+        if previous_step:
+            return self.step_should_run(part, previous_step)
+
+        return False
+
+    def add_step_run(self, part: pluginhandler.PluginHandler,
+                     step: steps.Step) -> None:
+        """Cache the fact that a given step has now run for the given part.
+
+        :param pluginhandler.PluginHandler part: Part in question.
+        :param steps.Step step: Step in question.
+        """
+        self._ensure_steps_run(part)
+        self._steps_run[part.name].add(step)
+
+    def step_has_run(self, part: pluginhandler.PluginHandler,
+                     step: steps.Step) -> bool:
+        """Determine if a given step of a given part has already run.
+
+        :param pluginhandler.PluginHandler part: Part in question.
+        :param steps.Step step: Step in question.
+        :return: Whether or not the step has run.
+        :rtype: bool
+        """
+        self._ensure_steps_run(part)
+        return step in self._steps_run[part.name]
+
+    def get_dirty_report(self, part: pluginhandler.PluginHandler,
+                         step: steps.Step) -> pluginhandler.DirtyReport:
+        """Obtain the dirty report for a given step of the given part.
+
+        :param pluginhandler.PluginHandler part: Part in question.
+        :param steps.Step step: Step in question.
+        :return: Dirty report (could be None)
+        :rtype: pluginhandler.DirtyReport
+        """
+        self._ensure_dirty_report(part, step)
+        return self._dirty_reports[part.name][step]
+
+    def clear_step(self, part: pluginhandler.PluginHandler,
+                   step: steps.Step) -> None:
+        """Clear the given step of the given part from the cache.
+
+        :param pluginhandler.PluginHandler part: Part in question.
+        :param steps.Step step: Step in question.
+
+        This function does nothing if the step wasn't cached.
+        """
+        if part.name in self._steps_run:
+            _remove_key(self._steps_run[part.name], step)
+            if not self._steps_run[part.name]:
+                _del_key(self._steps_run, part.name)
+        _del_key(self._dirty_reports[part.name], step)
+        if not self._dirty_reports[part.name]:
+            _del_key(self._dirty_reports, part.name)
+
+    def _ensure_steps_run(self, part: pluginhandler.PluginHandler) -> None:
+        if part.name not in self._steps_run:
+            self._steps_run[part.name] = _get_steps_run(part)
+
+    def _ensure_dirty_report(self, part: pluginhandler.PluginHandler,
+                             step: steps.Step) -> None:
+        if step not in self._dirty_reports[part.name]:
```

Comment:
how about returning early if it is in and bumping everything below one level in?
It would make it obvious that there is no `else` case below

Reply:
fwiw, this is the only scenario where I am mostly ok with multiple return scenarios, i.e.; prechecks.

Reply:
Done, actually was able to bring it back two stops.